### PR TITLE
sqlccl: move IMPORT to jobs registry

### PR DESCRIPTION
**Pull request by madelynnblue**
@andreimatei can you review the first commit and eval.go, planner.go, and server.go in the second commit?

And either David or Nikhil for the rest. Sorry it's a lot and no one else really knows the code. I tried to split it up into logical parts.

### CODE REVIEW

**Comment by knz** [comment]
Is there no way to have the distsql serialization read this data without going through these timestamp methods?

During evaluation it's still good to have them, to ensure that wrong times do not get evaluated.

---

Review status: 0 of 25 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22291#-:-L4bbiAG-c1pmrL_QbuJ:b-cfukwi)*

**Comment by dt** [comment]
Reviewed 11 of 11 files at r1, 10 of 15 files at r2.
Review status: 17 of 25 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/ccl/sqlccl/restore.go, line 808 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22291#-L4behzMQwGNGiU8Dj7q:-L4behzMQwGNGiU8Dj7r:b-o4p1h4) ([raw file](https://github.com/cockroachdb/cockroach/blob/54d096430c1b9ce1179a3ee4d66b89554b41d31b/pkg/ccl/sqlccl/restore.go#L808)):*
> ```Go
> 		if err := txn.Run(ctx, b); err != nil {
> 			if _, ok := errors.Cause(err).(*roachpb.ConditionFailedError); ok {
> 				return errors.New("table already exists")
> ```

Should we log the underlying error before we throw it out just incase some other source of ConditionFailed creeps in at some point? I see not Wrapf'ing in the interest of returning a prettier error, but if/when something else gets incorrectly transformed to table-exists, I could see it being confusing. /shrug.

---

*[pkg/sql/jobs/jobs.proto, line 80 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22291#-L4bcTNDltK-5GNvD_3d:-L4bcTNDltK-5GNvD_3e:bm56hb3) ([raw file](https://github.com/cockroachdb/cockroach/blob/54d096430c1b9ce1179a3ee4d66b89554b41d31b/pkg/sql/jobs/jobs.proto#L80)):*
> ```protobuf
>     int32 comment = 10;
>     // nullif needs to differentiate between not present and the empty string.
>     bool has_nullif = 11;
> ```

if you wanted to avoid the separate bool, could also make a message type that would be nullable, and then put the string inside that?

---

*[pkg/sql/sem/tree/eval.go, line 2189 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22291#-L4bboJifkXeomEsIS-K:-L4bboJifkXeomEsIS-L:b-hn2i3h) ([raw file](https://github.com/cockroachdb/cockroach/blob/5e9f236fd3020cace6c8bad5563d1502d45c14f9/pkg/sql/sem/tree/eval.go#L2189)):*
> ```Go
> // the evaluation context. The timestamp is guaranteed to be nonzero.
> func (ctx *EvalContext) GetStmtTimestamp() time.Time {
> 	// TODO(knz): a zero timestamp should never be read, even during
> ```

I assume these assertions were protecting from some sort of misuse... is it safe to just remove them?

---

*[pkg/testutils/lint/lint_test.go, line 948 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22291#-L4bYtRITYSKX8Zo0-Ht:-L4bYtRITYSKX8Zo0-Hu:bl5mptj) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce31446783c4da331374f94d111b4271c995b011/pkg/testutils/lint/lint_test.go#L948)):*
> ```Go
> 				stream.GrepNot("_fsm.go.*should not use dot imports"),
> 				// sql.planner is gonna stay unexported for now.
> 				stream.GrepNot("exported func NewInternalPlanner returns unexported type"),
> ```

This is still going to trip up editors that run lint file-by-file and such. I wonder if we should make an exported type that wraps a planner rather than converting all the internal callsites?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22291)*

**Comment by mjibson** [comment]
Yes, I've changed distsql to use the fields directly and restored the assertions.

---

Review status: 9 of 25 files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[pkg/ccl/sqlccl/restore.go, line 808 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22291#-L4behzMQwGNGiU8Dj7q:-L4bvqvkx299Cw95K8F6:b7l5gel) ([raw file](https://github.com/cockroachdb/cockroach/blob/54d096430c1b9ce1179a3ee4d66b89554b41d31b/pkg/ccl/sqlccl/restore.go#L808)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

Should we log the underlying error before we throw it out just incase some other source of ConditionFailed creeps in at some point? I see not Wrapf'ing in the interest of returning a prettier error, but if/when something else gets incorrectly transformed to table-exists, I could see it being confusing. /shrug.
</blockquote></details>

I'd prefer not to log anything because I think it will just add to the log spam and be of really little value. Since we are directly calling CPut here, we have some pretty well defined semantics about what a conditionfailederror means. If what you describe does happen (a user gets this error but the table doesn't exist), they will hopefully report it and we can figure out what changed.

---

*[pkg/sql/jobs/jobs.proto, line 80 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22291#-L4bcTNDltK-5GNvD_3d:-L4bxsVR0Q9tvaoLdk1F:begcc0j) ([raw file](https://github.com/cockroachdb/cockroach/blob/54d096430c1b9ce1179a3ee4d66b89554b41d31b/pkg/sql/jobs/jobs.proto#L80)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

if you wanted to avoid the separate bool, could also make a message type that would be nullable, and then put the string inside that?
</blockquote></details>

Since we are using proto3 nullable isn't an option. You can do this weird oneof thing with only a single field, but then you have to do Go type assertions to see what the type of the field is and it just adds a bunch of code complexity for a few proto byte savings. Definitely not worth it for this solution that is totally obvious how it works.

---

*[pkg/testutils/lint/lint_test.go, line 948 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22291#-L4bYtRITYSKX8Zo0-Ht:-L4bvGxbvSfoBvAYVpgL:b-g3y2l3) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce31446783c4da331374f94d111b4271c995b011/pkg/testutils/lint/lint_test.go#L948)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

This is still going to trip up editors that run lint file-by-file and such. I wonder if we should make an exported type that wraps a planner rather than converting all the internal callsites?
</blockquote></details>

I considered that. That would leave us with sql.planner, sql.PlanHookState, and a sql.ExportedPlanner that is always actually just an interface{} because of the package dependency stuff. I think it's better as is. We already do lots of not lint friendly stuff, like the fsm things one line above.

---

*[pkg/sql/sem/tree/eval.go, line 2189 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22291#-L4bboJifkXeomEsIS-K:-L4bvAYZ90Sh0lknm2yq:bjsk1e3) ([raw file](https://github.com/cockroachdb/cockroach/blob/5e9f236fd3020cace6c8bad5563d1502d45c14f9/pkg/sql/sem/tree/eval.go#L2189)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

I assume these assertions were protecting from some sort of misuse... is it safe to just remove them?
</blockquote></details>

I've changed the distsql serialization to not call these functions. The assertions have been restored.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22291#-:-L4by99SWDlC7X821PWk:brfs0v0)*

**Comment by dt** [comment]
Review status: 9 of 25 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[pkg/ccl/sqlccl/restore.go, line 808 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22291#-L4behzMQwGNGiU8Dj7q:-L4byquijuj7skLKeaJK:b4c9lyr) ([raw file](https://github.com/cockroachdb/cockroach/blob/54d096430c1b9ce1179a3ee4d66b89554b41d31b/pkg/ccl/sqlccl/restore.go#L808)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

I'd prefer not to log anything because I think it will just add to the log spam and be of really little value. Since we are directly calling CPut here, we have some pretty well defined semantics about what a conditionfailederror means. If what you describe does happen (a user gets this error but the table doesn't exist), they will hopefully report it and we can figure out what changed.
</blockquote></details>

:thumbsup: 

---

*[pkg/sql/jobs/jobs.proto, line 80 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22291#-L4bcTNDltK-5GNvD_3d:-L4byVZqAo8Fj_dL4q45:bfs8kiw) ([raw file](https://github.com/cockroachdb/cockroach/blob/54d096430c1b9ce1179a3ee4d66b89554b41d31b/pkg/sql/jobs/jobs.proto#L80)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

Since we are using proto3 nullable isn't an option. You can do this weird oneof thing with only a single field, but then you have to do Go type assertions to see what the type of the field is and it just adds a bunch of code complexity for a few proto byte savings. Definitely not worth it for this solution that is totally obvious how it works.
</blockquote></details>

 fwiw, I was just picturing message type with a single string field in it, and then a (nullable) field of that message here -- no oneof or anything  fancy -- but I don't feel strongly.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22291)*

**Comment by mjibson** [comment]
Review status: 9 of 25 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[pkg/sql/jobs/jobs.proto, line 80 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22291#-L4bcTNDltK-5GNvD_3d:-L4c-H-kx2uKWmsBrJ_Y:bc8oibg) ([raw file](https://github.com/cockroachdb/cockroach/blob/54d096430c1b9ce1179a3ee4d66b89554b41d31b/pkg/sql/jobs/jobs.proto#L80)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

 fwiw, I was just picturing message type with a single string field in it, and then a (nullable) field of that message here -- no oneof or anything  fancy -- but I don't feel strongly.
</blockquote></details>

There's no nullable anymore in proto3.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22291)*

**Comment by andreimatei** [comment]
:lgtm:

---

Review status: 9 of 25 files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---

*[pkg/sql/planner.go, line 229 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/22291#-L4c-8ONl65idtnhE6bh:-L4c-8ONl65idtnhE6bi:b-kecnxg) ([raw file](https://github.com/cockroachdb/cockroach/blob/a63410211489a8f55e6bb355cd65f5768e95beab/pkg/sql/planner.go#L229)):*
> ```Go
> 		ts = txn.Proto().OrigTimestamp.GoTime()
> 	}
> 	// If txn is nil, none of the timestamp fields of the eval ctx will be set
> ```

please put this comment as an arg comment on the function definition

---

*[pkg/sql/jobs/registry.go, line 78 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/22291#-L4bzYVPf7zDIstsjn2B:-L4bzYVPf7zDIstsjn2C:b-oi549r) ([raw file](https://github.com/cockroachdb/cockroach/blob/a63410211489a8f55e6bb355cd65f5768e95beab/pkg/sql/jobs/registry.go#L78)):*
> ```Go
> // stuff that PlanHookState exports. One other choice is to merge this package
> // back into the sql package. There's maybe a better way that I'm unaware of.
> type planHookMaker func(opName, user string) (interface{}, func())
> ```

please spell out what this returns behind the interface, and what the returned function is.

---

*[pkg/testutils/lint/lint_test.go, line 948 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22291#-L4bYtRITYSKX8Zo0-Ht:-L4byspZF6PIuD67xhOP:b-hv0u4m) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce31446783c4da331374f94d111b4271c995b011/pkg/testutils/lint/lint_test.go#L948)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

I considered that. That would leave us with sql.planner, sql.PlanHookState, and a sql.ExportedPlanner that is always actually just an interface{} because of the package dependency stuff. I think it's better as is. We already do lots of not lint friendly stuff, like the fsm things one line above.
</blockquote></details>

how about an exported version of `newInternalPlanner` that returns `interface{}`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22291#-:-L4c-kByPUSun2oZ5upm:bnfp4nl)*

**Comment by dt** [comment]
Review status: 9 of 25 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[pkg/sql/jobs/jobs.proto, line 80 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22291#-L4bcTNDltK-5GNvD_3d:-L4cCx2ypqAIqhQecGgb:bvwhuur) ([raw file](https://github.com/cockroachdb/cockroach/blob/54d096430c1b9ce1179a3ee4d66b89554b41d31b/pkg/sql/jobs/jobs.proto#L80)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

There's no nullable anymore in proto3.
</blockquote></details>

Messages are nullable I believe — just primitives that aren’t.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22291)*

File: pkg/sql/jobs/registry.go [hunk 1]
```diff
@@ -71,26 +68,35 @@ type Registry struct {
...
 }
 
-// MakeRegistry creates a new Registry.
+// TODO(mjibson): Can we do something to avoid passing an interface{} here
+// that must be type casted in a Resumer? It cannot be done here because
+// PlanHookState lives in the sql package, which would create a dependency
+// cycle if listed here. Furthermore, moving PlanHookState into a common
+// subpackage like sqlbase is difficult because of the amount of sql-only
+// stuff that PlanHookState exports. One other choice is to merge this package
+// back into the sql package. There's maybe a better way that I'm unaware of.
```
Comment (benesch) [inline pkg/sql/jobs/registry.go]: Let me sleep on this, if you don't mind!

**Comment by mjibson** [comment]
Review status: 9 of 25 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[pkg/sql/planner.go, line 229 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/22291#-L4c-8ONl65idtnhE6bh:-L4dlikAyQp0_oUl1gOK:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a63410211489a8f55e6bb355cd65f5768e95beab/pkg/sql/planner.go#L229)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

please put this comment as an arg comment on the function definition
</blockquote></details>

Done.

---

*[pkg/sql/jobs/jobs.proto, line 80 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22291#-L4bcTNDltK-5GNvD_3d:-L4dzxDmAMz2QtV-ywgD:bii2jon) ([raw file](https://github.com/cockroachdb/cockroach/blob/54d096430c1b9ce1179a3ee4d66b89554b41d31b/pkg/sql/jobs/jobs.proto#L80)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

Messages are nullable I believe — just primitives that aren’t.
</blockquote></details>

Ah yes. Done.

---

*[pkg/sql/jobs/registry.go, line 78 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/22291#-L4bzYVPf7zDIstsjn2B:-L4dmR2wQALVsJDzkfLv:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a63410211489a8f55e6bb355cd65f5768e95beab/pkg/sql/jobs/registry.go#L78)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

please spell out what this returns behind the interface, and what the returned function is.
</blockquote></details>

Done.

---

*[pkg/testutils/lint/lint_test.go, line 948 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22291#-L4bYtRITYSKX8Zo0-Ht:-L4dyo7CTVHiWtu3-rlJ:b-duzahh) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce31446783c4da331374f94d111b4271c995b011/pkg/testutils/lint/lint_test.go#L948)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

how about an exported version of `newInternalPlanner` that returns `interface{}`?
</blockquote></details>

Yes! Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22291)*

**Comment by dt** [comment]
Review status: 0 of 16 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[pkg/sql/jobs/jobs.proto, line 80 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22291#-L4bcTNDltK-5GNvD_3d:-L4fbZiU8MY5Fb-jD4-y:bn1wz0t) ([raw file](https://github.com/cockroachdb/cockroach/blob/54d096430c1b9ce1179a3ee4d66b89554b41d31b/pkg/sql/jobs/jobs.proto#L80)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

Ah yes. Done.
</blockquote></details>

:thumbsup:

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22291)*

**Comment by mjibson** [comment]
@andreimatei Can you take a look at the current NewInternalPlanner and distsql timestamp stuff in api.go since it changed a bit? Will merge if you think it's ok.

**Comment by andreimatei** [comment]
:lgtm: but pls coordinate with @knz on #22440, which removes `EvalCtx.ClusterTimestamp`

---

Review status: 0 of 17 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22291#-:-L4iBYQuqh8YWxb2bDOk:bi9dpvb)*

**Comment by mjibson** [comment]
@knz Do you want me to merge this PR or should I wait for yours? I'd prefer to do this one first since I don't want to have to undo some work you are doing in that other PR.

**Comment by knz** [comment]
Please merge this first, I'll adjust the other one accordingly. Thanks

**Pull request merged**